File: src/io/write.ts
// node:fs/promises for mkdir/rename — Bun has no built-in atomic rename.
import { mkdir, rename } from "node:fs/promises";
import { dirname, join } from "node:path";

// Atomic write (tmp + rename). Deliberately NO before-write backup: v2 never
// snapshots implicitly (spec requirement 3); `menv backup` is the only backup.
export async function writeFileAtomic(
  root: string,
  rel: string,
  content: string | Uint8Array,
): Promise<string> {
  const abs = join(root, rel);
  await mkdir(dirname(abs), { recursive: true });
  const tmp = `${abs}.menv-tmp`;
  await Bun.write(tmp, content);
  await rename(tmp, abs);
  return rel;
}

File: go/internal/io/write.go
package io

import (
	"fmt"
	"os"
	"path/filepath"
)

// WriteFileAtomic writes content to root/rel using a tmp+rename sequence so
// the target file is never in a partially-written state. Parent directories are
// created as needed. There is NO implicit backup — only `menv backup` snapshots.
func WriteFileAtomic(root, rel string, content []byte) error {
	abs := filepath.Join(root, rel)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		return fmt.Errorf("creating parent dirs for %s: %w", rel, err)
	}
	tmp := abs + ".menv-tmp"
	if err := os.WriteFile(tmp, content, 0o644); err != nil {
		return fmt.Errorf("writing tmp file %s: %w", tmp, err)
	}
	if err := os.Rename(tmp, abs); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("renaming %s → %s: %w", tmp, abs, err)
	}
	return nil
}
